### PR TITLE
[RFC] mac80211/hostapd: add support for 256-QAM on 802.11n

### PR DESCRIPTION
**Pull request by dengqf6**
Some vendors allow VHT (256-QAM) rate on 802.11n, but mac80211 does not allow it (yet).
Add a hack to let mac80211 allow VHT

Tested on IPQ4019 and mvebu/mwlwifi

mac80211 patch was from @Ansuel

### CODE REVIEW

**Comment by dengqf6** [comment]
@Ansuel
Your patch did not add 'vendor_vht' option to hostapd, which is required by 256-QAM

**Comment by Ansuel** [comment]
Anyway can iwinfo display the use of 256-qam in any way?

**Comment by dengqf6** [comment]
If the client supports 256-QAM, Wireless status page will show VHT rate

**Comment by psyborg55** [comment]
how do you configure AP interface? luci or cli ? because luci have no 80MHz support on 2.4band. tried default settings here (20MHz) and 40MHz both with and without noscan, no rates shown other than 144/300Mbps once client connected. added vendor_vht '1' to wifi options and patched mac80211 on client side as well but the rates do not change

**Comment by dengqf6** [comment]
@psyborg55
via uci/luci
check if `vendor_vht=1` is in `/var/run/hostapd-phyX.conf`
Not every device supports VHT on 2.4GHz. on some Android phones you have to root the phone and edit the driver config to enable it.

**Comment by psyborg55** [comment]
yes, vendor_vht=1 is in hostapd conf. if i set htmode 'VHT20' in wireless config then i notice there is no vht_oper_chwidth=1 in hostapd conf. after setting VHT40 or VHT80 this parameter appears in hostapd. after this i cannot open wireless editing from luci; typeerror returns: vals is undefined
should i add vendor_vht parameter to client side as well? it is linux laptop (k4.9.172), driver info:

ubuntu@ubuntu:~/backports-5.4-rc1-test4$ iw list
Wiphy phy0
	max # scan SSIDs: 16
	max scan IEs length: 195 bytes
	max # sched scan SSIDs: 0
	max # match sets: 0
	max # scan plans: 1
	max scan plan interval: -1
	max scan plan iterations: 0
	Retry short limit: 7
	Retry long limit: 4
	Coverage class: 0 (up to 0m)
	Device supports RSN-IBSS.
	Device supports AP-side u-APSD.
	Supported Ciphers:
		* WEP40 (00-0f-ac:1)
		* WEP104 (00-0f-ac:5)
		* TKIP (00-0f-ac:2)
		* CCMP-128 (00-0f-ac:4)
		* CMAC (00-0f-ac:6)
		* CMAC-256 (00-0f-ac:13)
		* GMAC-128 (00-0f-ac:11)
		* GMAC-256 (00-0f-ac:12)
	Available Antennas: TX 0x3 RX 0x3
	Configured Antennas: TX 0x3 RX 0x3
	Supported interface modes:
		 * managed
		 * AP
		 * monitor
	Band 1:
		Capabilities: 0x19ef
			RX LDPC
			HT20/HT40
			SM Power Save disabled
			RX HT20 SGI
			RX HT40 SGI
			TX STBC
			RX STBC 1-stream
			Max AMSDU length: 7935 bytes
			DSSS/CCK HT40
		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
		Minimum RX AMPDU time spacing: 8 usec (0x06)
		HT TX/RX MCS rate indexes supported: 0-15
		VHT Capabilities (0x338001b2):
			Max MPDU length: 11454
			Supported Channel Width: neither 160 nor 80+80
			RX LDPC
			short GI (80 MHz)
			TX STBC
			RX antenna pattern consistency
			TX antenna pattern consistency
		VHT RX MCS set:
			1 streams: MCS 0-9
			2 streams: MCS 0-9
			3 streams: not supported
			4 streams: not supported
			5 streams: not supported
			6 streams: not supported
			7 streams: not supported
			8 streams: not supported
		VHT RX highest supported: 0 Mbps
		VHT TX MCS set:
			1 streams: MCS 0-9
			2 streams: MCS 0-9
			3 streams: not supported
			4 streams: not supported
			5 streams: not supported
			6 streams: not supported
			7 streams: not supported
			8 streams: not supported
		VHT TX highest supported: 0 Mbps
		Bitrates (non-HT):
			* 1.0 Mbps
			* 2.0 Mbps (short preamble supported)
			* 5.5 Mbps (short preamble supported)
			* 11.0 Mbps (short preamble supported)
			* 6.0 Mbps
			* 9.0 Mbps
			* 12.0 Mbps
			* 18.0 Mbps
			* 24.0 Mbps
			* 36.0 Mbps
			* 48.0 Mbps
			* 54.0 Mbps
		Frequencies:
			* 2377 MHz [250] (30.0 dBm)
			* 2382 MHz [251] (30.0 dBm)
			* 2387 MHz [252] (30.0 dBm)
			* 2392 MHz [253] (30.0 dBm)
			* 2397 MHz [254] (30.0 dBm)
			* 2402 MHz [255] (30.0 dBm)
			* 2407 MHz [256] (30.0 dBm)
			* 2412 MHz [1] (30.0 dBm)
			* 2417 MHz [2] (30.0 dBm)
			* 2422 MHz [3] (30.0 dBm)
			* 2427 MHz [4] (30.0 dBm)
			* 2432 MHz [5] (30.0 dBm)
			* 2437 MHz [6] (30.0 dBm)
			* 2442 MHz [7] (30.0 dBm)
			* 2447 MHz [8] (30.0 dBm)
			* 2452 MHz [9] (30.0 dBm)
			* 2457 MHz [10] (30.0 dBm)
			* 2462 MHz [11] (30.0 dBm)
			* 2467 MHz [12] (30.0 dBm)
			* 2472 MHz [13] (30.0 dBm)
			* 2484 MHz [14] (30.0 dBm)
			* 2537 MHz [25] (30.0 dBm)
	Band 2:
		Capabilities: 0x19ef
			RX LDPC
			HT20/HT40
			SM Power Save disabled
			RX HT20 SGI
			RX HT40 SGI
			TX STBC
			RX STBC 1-stream
			Max AMSDU length: 7935 bytes
			DSSS/CCK HT40
		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
		Minimum RX AMPDU time spacing: 8 usec (0x06)
		HT TX/RX MCS rate indexes supported: 0-15
		VHT Capabilities (0x338001b2):
			Max MPDU length: 11454
			Supported Channel Width: neither 160 nor 80+80
			RX LDPC
			short GI (80 MHz)
			TX STBC
			RX antenna pattern consistency
			TX antenna pattern consistency
		VHT RX MCS set:
			1 streams: MCS 0-9
			2 streams: MCS 0-9
			3 streams: not supported
			4 streams: not supported
			5 streams: not supported
			6 streams: not supported
			7 streams: not supported
			8 streams: not supported
		VHT RX highest supported: 0 Mbps
		VHT TX MCS set:
			1 streams: MCS 0-9
			2 streams: MCS 0-9
			3 streams: not supported
			4 streams: not supported
			5 streams: not supported
			6 streams: not supported
			7 streams: not supported
			8 streams: not supported
		VHT TX highest supported: 0 Mbps
		Bitrates (non-HT):
			* 6.0 Mbps
			* 9.0 Mbps
			* 12.0 Mbps
			* 18.0 Mbps
			* 24.0 Mbps
			* 36.0 Mbps
			* 48.0 Mbps
			* 54.0 Mbps
		Frequencies:
			* 5180 MHz [36] (30.0 dBm)
			* 5200 MHz [40] (30.0 dBm)
			* 5220 MHz [44] (30.0 dBm)
			* 5240 MHz [48] (30.0 dBm)
			* 5260 MHz [52] (30.0 dBm)
			* 5280 MHz [56] (30.0 dBm)
			* 5300 MHz [60] (30.0 dBm)
			* 5320 MHz [64] (30.0 dBm)
			* 5500 MHz [100] (30.0 dBm)
			* 5520 MHz [104] (30.0 dBm)
			* 5540 MHz [108] (30.0 dBm)
			* 5560 MHz [112] (30.0 dBm)
			* 5580 MHz [116] (30.0 dBm)
			* 5600 MHz [120] (30.0 dBm)
			* 5620 MHz [124] (30.0 dBm)
			* 5640 MHz [128] (30.0 dBm)
			* 5660 MHz [132] (30.0 dBm)
			* 5680 MHz [136] (30.0 dBm)
			* 5700 MHz [140] (30.0 dBm)
			* 5720 MHz [144] (30.0 dBm)
			* 5745 MHz [149] (30.0 dBm)
			* 5765 MHz [153] (30.0 dBm)
			* 5785 MHz [157] (30.0 dBm)
			* 5805 MHz [161] (30.0 dBm)
			* 5825 MHz [165] (30.0 dBm)
			* 5845 MHz [169] (30.0 dBm)
			* 5865 MHz [173] (30.0 dBm)
	Supported commands:
		 * new_interface
		 * set_interface
		 * new_key
		 * start_ap
		 * new_station
		 * set_bss
		 * authenticate
		 * associate
		 * deauthenticate
		 * disassociate
		 * join_ibss
		 * remain_on_channel
		 * set_tx_bitrate_mask
		 * frame
		 * frame_wait_cancel
		 * set_wiphy_netns
		 * set_channel
		 * set_wds_peer
		 * probe_client
		 * set_noack_map
		 * register_beacons
		 * start_p2p_device
		 * set_mcast_rate
		 * connect
		 * disconnect
		 * channel_switch
		 * set_qos_map
		 * set_multicast_to_unicast
	Supported TX frame types:
		 * IBSS: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * managed: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP/VLAN: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * mesh point: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-client: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-GO: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-device: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
	Supported RX frame types:
		 * IBSS: 0x40 0xb0 0xc0 0xd0
		 * managed: 0x40 0xd0
		 * AP: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * AP/VLAN: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * mesh point: 0xb0 0xc0 0xd0
		 * P2P-client: 0x40 0xd0
		 * P2P-GO: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * P2P-device: 0x40 0xd0
	software interface modes (can always be added):
		 * monitor
	valid interface combinations:
		 * #{ AP } <= 8, #{ managed } <= 1,
		   total <= 8, #channels <= 1, STA/AP BI must match
	HT Capability overrides:
		 * MCS: ff ff ff ff ff ff ff ff ff ff
		 * maximum A-MSDU length
		 * supported channel width
		 * short GI for 40 MHz
		 * max A-MPDU length exponent
		 * min MPDU start spacing
	Device supports TX status socket option.
	Device supports HT-IBSS.
	Device supports SAE with AUTHENTICATE command
	Device supports scan flush.
	Device supports AP scan.
	Device supports per-vif TX power setting
	Driver supports full state transitions for AP/GO clients
	Driver supports a userspace MPM
	Driver/device bandwidth changes during BSS lifetime (AP/GO mode)
	Device supports static SMPS
	Device supports configuring vdev MAC-addr on create.
	Device supports VHT-IBSS.
ubuntu@ubuntu:~/backports-5.4-rc1-test4$

File: package/kernel/mac80211/patches/subsys/600-mac80211-allow-vht-on-2g.patch [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ 		have_80mhz = true;
+--- a/net/mac80211/util.c
++++ b/net/mac80211/util.c
+@@ -1769,7 +1769,8 @@ static int ieee80211_build_preq_ies_band
+ 	/* Check if any channel in this sband supports at least 80 MHz */
+ 	for (i = 0; i < sband->n_channels; i++) {
+ 		if (sband->channels[i].flags & (IEEE80211_CHAN_DISABLED |
+-						IEEE80211_CHAN_NO_80MHZ))
++						IEEE80211_CHAN_NO_80MHZ) &
++						(sband->band != NL80211_BAND_2GHZ))
```
Comment (oldium) [inline package/kernel/mac80211/patches/subsys/600-mac80211-allow-vht-on-2g.patch]: Just spotted this - bitwise AND with a boolean value? This does not look correct. Same few lines below. Is this intentional?

Comment (dengqf6) [reply package/kernel/mac80211/patches/subsys/600-mac80211-allow-vht-on-2g.patch]: @Ansuel

Comment (Ansuel) [reply package/kernel/mac80211/patches/subsys/600-mac80211-allow-vht-on-2g.patch]:
```
if ((sband->channels[i].flags & (IEEE80211_CHAN_DISABLED |
						IEEE80211_CHAN_NO_80MHZ)) &&
						(sband->band != NL80211_BAND_2GHZ))
```

@oldium yep should be totally wrong... this should be the correct condition right?

**Comment by Ansuel** [comment]
@dengqf6 tested with my s8
Annotazione 2019-11-22 022309

**Comment by CodeFetch** [comment]
@dengqf6 Why can't the kernel part be upstreamed?

**Comment by Ansuel** [comment]
@CodeFetch this is totally outside specs....

**Comment by dengqf6** [comment]
fyi: the ath10k part was upstream before, but got reverted
ath10k-ct still has it though

**Comment by adrianschmutzler** [comment]
@dengqf6 There doesn't seem to be interest by any of the committers, and you called it a "hack" yourself initially. Are you fine with me closing this, as last relevant comment has been half a year ago?

**Comment by dengqf6** [comment]
It would be nice to have this feature because a lot of routers, phones and NICs support it too..

**Comment by adrianschmutzler** [comment]
> It would be nice to have this feature because a lot of routers, phones and NICs support it too..

Might be, but so far almost nobody was interested in the proposal and I don't think this will change just because we wait another 6 months. You might try to increase your audience by resubmitting to the mailing list.

**Comment by psyborg55** [comment]
the PR seems to be incomplete/didn't work.i think it needs more changes

**Comment by dengqf6** [comment]
> the PR seems to be incomplete/didn't work.i think it needs more changes

Did you patch the client mac80211?

**Comment by psyborg55** [comment]
of course. what i don't see here is lua support for interface setting. and mac80211 change looks to simple for this to work. unless i am mistaken about hw support and compex card cannot work in 2.4VHT?

**Comment by Ansuel** [comment]
there is no setting for this kind of things... this just works if the device supports it or not... (I posted a screen of this working)

**Comment by oldium** [comment]
Doesn't this enable VHT on 2.4GHz band for all devices - also those without actual support?

**Comment by dengqf6** [comment]
> Doesn't this enable VHT on 2.4GHz band for all devices - also those without actual support?

This does not just enable VHT unconditionally - it reads `vendor_vht` from uci

**Comment by dengqf6** [comment]
@Borromini It should work on DIR860L too

**Comment by DavideFioravanti** [comment]
I can confirm that it works on mt7615 with some minor modification to this patch: https://github.com/openwrt/mt76/pull/333

This is my testing configuration:
```
config wifi-device 'radio0'
        option type 'mac80211'
        option path '1e140000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0'
        option hwmode '11g'
        option vendor_vht '1'
        option htmode 'HT40'
        option noscan '1'
        option country 'IT'
        option channel '13'
```
and this is the result:
image

**Comment by Ansuel** [comment]
Since we have at least 2 vendor that supports this (ath10k and mt)  and our version ath10k-ct already have this enabled in the code,  I really think this should be pushed, even if outside spec. Not using this would cause performance regression than the stock firmware and I can't really see any problem with using this.

**Comment by CodeFetch** [comment]
It's not more controversial than Broadcom's "frameburst" or narrow channels which are already supported by OpenWrt... Thus it should just be merged.

**Comment by tiagogaspar8** [comment]
Is this getting merged upstream? I saw some patches but i couldn't see if they were rejected or the person gave up.
Either way I agree with this patch, it's a speed improvement, doesn't cause problems, so I see no issue.

**Comment by Ansuel** [comment]
Can't merged upstream. (It's really a special implementation of the OEM...
Anyway since this is not that big, I think we can just keep it as a hack
and merge to openwrt)

Il Gio 26 Nov 2020, 15:32 Tiago Gaspar <notifications@github.com> ha
scritto:

**Comment by tiagogaspar8** [comment]
Yeah I agree, it it's just a new feature, not a braking change.
Is anything missing?

**Comment by Vladdrako** [comment]
Tried on mir3g. I have only 1Mbps of real speed. (Tested with speedtest.) In LuCi I see only (and only) this:
```
6.0 Mbit/s, 20 МГц
6.5 Mbit/s, 20 МГц, VHT-MCS 0, VHT-NSS 1
```
So, 40MHz is not using.
<details>
  <summary>config</summary>
<pre>
config wifi-device 'radio0'
	option type 'mac80211'
	option hwmode '11g'
	option path '1e140000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0'
	option htmode 'HT40'
	option country 'VE'
	option txpower '26'
	option channel '6'
	option noscan '1'
	option legacy_rates '1'
	option vendor_vht '1'
</pre>
</details>
<details>
  <summary>iw phy0 info</summary>
<pre>
Wiphy phy0
        wiphy index: 0
        max # scan SSIDs: 4
        max scan IEs length: 2243 bytes
        max # sched scan SSIDs: 0
        max # match sets: 0
        max # scan plans: 1
        max scan plan interval: -1
        max scan plan iterations: 0
        Retry short limit: 7
        Retry long limit: 4
        Coverage class: 0 (up to 0m)
        Device supports RSN-IBSS.
        Device supports AP-side u-APSD.
        Device supports T-DLS.
        Supported Ciphers:
                * WEP40 (00-0f-ac:1)
                * WEP104 (00-0f-ac:5)
                * TKIP (00-0f-ac:2)
                * CCMP-128 (00-0f-ac:4)
                * CCMP-256 (00-0f-ac:10)
                * GCMP-128 (00-0f-ac:8)
                * GCMP-256 (00-0f-ac:9)
                * CMAC (00-0f-ac:6)
                * CMAC-256 (00-0f-ac:13)
                * GMAC-128 (00-0f-ac:11)
                * GMAC-256 (00-0f-ac:12)
        Available Antennas: TX 0x3 RX 0x3
        Configured Antennas: TX 0x3 RX 0x3
        Supported interface modes:
                 * IBSS
                 * managed
                 * AP
                 * AP/VLAN
                 * monitor
                 * mesh point
                 * P2P-client
                 * P2P-GO
        Band 1:
                Capabilities: 0x1fe
                        HT20/HT40
                        SM Power Save disabled
                        RX Greenfield
                        RX HT20 SGI
                        RX HT40 SGI
                        TX STBC
                        RX STBC 1-stream
                        Max AMSDU length: 3839 bytes
                        No DSSS/CCK HT40
                Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
                Minimum RX AMPDU time spacing: No restriction (0x00)
                HT TX/RX MCS rate indexes supported: 0-15
                VHT Capabilities (0x318001b0):
                        Max MPDU length: 3895
                        Supported Channel Width: neither 160 nor 80+80
                        RX LDPC
                        short GI (80 MHz)
                        TX STBC
                        RX antenna pattern consistency
                        TX antenna pattern consistency
                VHT RX MCS set:
                        1 streams: MCS 0-9
                        2 streams: MCS 0-9
                        3 streams: not supported
                        4 streams: not supported
                        5 streams: not supported
                        6 streams: not supported
                        7 streams: not supported
                        8 streams: not supported
                VHT RX highest supported: 0 Mbps
                VHT TX MCS set:
                        1 streams: MCS 0-9
                        2 streams: MCS 0-9
                        3 streams: not supported
                        4 streams: not supported
                        5 streams: not supported
                        6 streams: not supported
                        7 streams: not supported
                        8 streams: not supported
                VHT TX highest supported: 0 Mbps
                Bitrates (non-HT):
                        * 1.0 Mbps (short preamble supported)
                        * 2.0 Mbps (short preamble supported)
                        * 5.5 Mbps (short preamble supported)
                        * 11.0 Mbps (short preamble supported)
                        * 6.0 Mbps
                        * 9.0 Mbps
                        * 12.0 Mbps
                        * 18.0 Mbps
                        * 24.0 Mbps
                        * 36.0 Mbps
                        * 48.0 Mbps
                        * 54.0 Mbps
                Frequencies:
                        * 2412 MHz [1] (26.0 dBm)
                        * 2417 MHz [2] (26.0 dBm)
                        * 2422 MHz [3] (26.0 dBm)
                        * 2427 MHz [4] (26.0 dBm)
                        * 2432 MHz [5] (26.0 dBm)
                        * 2437 MHz [6] (26.0 dBm)
                        * 2442 MHz [7] (26.0 dBm)
                        * 2447 MHz [8] (26.0 dBm)
                        * 2452 MHz [9] (26.0 dBm)
                        * 2457 MHz [10] (26.0 dBm)
                        * 2462 MHz [11] (26.0 dBm)
                        * 2467 MHz [12] (26.0 dBm)
                        * 2472 MHz [13] (26.0 dBm)
                        * 2484 MHz [14] (disabled)
        Supported commands:
                 * new_interface
                 * set_interface
                 * new_key
                 * start_ap
                 * new_station
                 * new_mpath
                 * set_mesh_config
                 * set_bss
                 * authenticate
                 * associate
                 * deauthenticate
                 * disassociate
                 * join_ibss
                 * join_mesh
                 * remain_on_channel
                 * set_tx_bitrate_mask
                 * frame
                 * frame_wait_cancel
                 * set_wiphy_netns
                 * set_channel
                 * set_wds_peer
                 * tdls_mgmt
                 * tdls_oper
                 * probe_client
                 * set_noack_map
                 * register_beacons
                 * start_p2p_device
                 * set_mcast_rate
                 * connect
                 * disconnect
                 * channel_switch
                 * set_qos_map
                 * set_multicast_to_unicast
        Supported TX frame types:
                 * IBSS: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * managed: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * AP: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * AP/VLAN: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * mesh point: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * P2P-client: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * P2P-GO: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
                 * P2P-device: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
        Supported RX frame types:
                 * IBSS: 0x40 0xb0 0xc0 0xd0
                 * managed: 0x40 0xb0 0xd0
                 * AP: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
                 * AP/VLAN: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
                 * mesh point: 0xb0 0xc0 0xd0
                 * P2P-client: 0x40 0xd0
                 * P2P-GO: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
                 * P2P-device: 0x40 0xd0
        software interface modes (can always be added):
                 * AP/VLAN
                 * monitor
        valid interface combinations:
                 * #{ IBSS } <= 1, #{ managed, AP, mesh point, P2P-client, P2P-GO } <= 4,
                   total <= 4, #channels <= 1, STA/AP BI must match
        HT Capability overrides:
                 * MCS: ff ff ff ff ff ff ff ff ff ff
                 * maximum A-MSDU length
                 * supported channel width
                 * short GI for 40 MHz
                 * max A-MPDU length exponent
                 * min MPDU start spacing
        Device supports TX status socket option.
        Device supports HT-IBSS.
        Device supports SAE with AUTHENTICATE command
        Device supports low priority scan.
        Device supports scan flush.
        Device supports AP scan.
        Device supports per-vif TX power setting
        Driver supports full state transitions for AP/GO clients
        Driver supports a userspace MPM
        Device supports active monitor (which will ACK incoming frames)
        Device supports configuring vdev MAC-addr on create.
        Supported extended features:
                * [ RRM ]: RRM
                * [ CQM_RSSI_LIST ]: multiple CQM_RSSI_THOLD records
                * [ CONTROL_PORT_OVER_NL80211 ]: control port over nl80211
                * [ TXQS ]: FQ-CoDel-enabled intermediate TXQs
                * [ AIRTIME_FAIRNESS ]: airtime fairness scheduling
                * [ AQL ]: Airtime Queue Limits (AQL)
                * [ SCAN_RANDOM_SN ]: use random sequence numbers in scans
                * [ SCAN_MIN_PREQ_CONTENT ]: use probe request with only rate IEs in scans
                * [ CONTROL_PORT_NO_PREAUTH ]: disable pre-auth over nl80211 control port support
                * [ DEL_IBSS_STA ]: deletion of IBSS station support
                * [ SCAN_FREQ_KHZ ]: scan on kHz frequency support
                * [ CONTROL_PORT_OVER_NL80211_TX_STATUS ]: tx status for nl80211 control port support
</pre>
</details>
So, mir3g is incompatible with this "hack" or I'm making an error somewhere? I built mt76 driver from the latest commit and hostapd/mac80211 with all necessary patches.

**Comment by adschm** [comment]
So, another 6 months are over, and still nobody interested in merging, or has at least shown any interest in bringing this forward. Thus, I'm finally considering this PR as "rejected". Sorry.

**Comment by tiagogaspar8** [comment]
Well, this is disappointing........

**Comment by Ansuel** [comment]
@adschm anyway the fun purt about this feature is the fact that mwlwifi driver and ath10k-ct driver already have this enabled (support in the code by default and enabled in the oem firmware). So it's really the mac80211 part that should be enabled.

**Comment by adschm** [comment]
@Ansuel There is no point in arguing with me about this one. I've closed this due to lack of response in slightly more than 1 year, not because I have a substantial opinion about the proposal itself.

**Comment by DavideFioravanti** [comment]
@adschm Even if I wish I'd seen this feature implemented, I respect your decision. 
I am going to ask the following question just to understand how to behave and maybe save you some time in the future... I come in peace :smile: 

What prevented this PR from merging in the end? Too few devs involved?
At least 4 of 5 people (including me) tested correctly this optional feature and giving the fact that in the worst case scenario it wouldn't do anything I don't know what was the problem about it. 
I am following this other PR (https://github.com/openwrt/openwrt/pull/3498) and I understand that a critical change like that needs a lot caution and lot of reviewers, but I think that this is a very different story...
I know that 4 or 5 devs is very very far from "a lot", but for a niche feature like this, probably can be adequate in my opinion. 
However If you think that this is not an adequate number of devs... well, case closed. :wink: 

P.s. I repeat: I am asking this with the best intentions and, above all, I am very grateful for your work in this community. Thanks. :wink: 

P.p.s Further considerations: I think that this feature didn't gain a lot of attention because it can have benefits in very limited use cases and probably because most of us (including me) avoids 2.4GHz channels whenever possible... but this is not always possible eg: multiple dual band devices used as repeaters in chain (wds). To not bottleneck the backhaul you can't use the same band also for the next segment especially if you have ETH clients. In this case, even 20-50 Mbit/s of the 200Mbit/s maximum theoretical gain is a valuable gain (probably it's a 10% increase over the non VHT 2.4GHz).
I implemented it before the 3-band devices became cheaper, but to be honest, I never did proper testing to calculate the gains. If some benchmarks can help to save this PR I could rebuild my old network in the following days/weeks using both ath10k and mt76 devices :)

**Comment by adschm** [comment]
> What prevented this PR from merging in the end?

The fact that nobody did it.

This is a feature, so adding it and maintaining it will require time (and expertise) for review and adjustments. Somebody of the decision makers has to provide that. In a commercial project a product manager would decide and probably calculate cost vs. benefit; in OpenWrt and other FOSS, there is no project manager. The decision is effectively made by whether one of the committer becomes interested or not.
This process obviously has much more entropy: Only a subset will be able to understand/review your changes, and among those there has to be somebody who has the time and interest and cares enough to give it sufficient priority. This essentially explains why some contributions are "taken" and others aren't. This might seem unfair to some authors that aren't picked, but it is objectively fair.

Coming back from the process to the subject:
This PR was opened 12 months ago, and in that entire time no committer even left a comment about the content. There is no reason to expect this will change in the next 12 months. So, the logical consequence is closing. Having piles of dead PRs won't help the development process, but just make organizing it harder.
(If you have read the history, you will be aware that I actually wanted to close this after the usual waiting time of 6 months (which I typically apply in these cases), but gave it another 6 months, in which nothing changed.)

Some people arrive here with the feeling that we _have to_ accept a change unless we can argument against it. But we are only spending our free time here, and can only accept a certain amount of stuff. So, the process is more like a _default-no_ on features nobody gets interested in, than the default-yes some people seem to expect.

So, what now?

You may just keep these changes locally and build yourself. Beyond that, there are the packages feeds. If not accepted into main repo, you in principle might try to package your changes and bring them to the packages feed, so people can download them _if_ they want them. This will also include more relaxed criteria for the merge, as it's really optional there, and we don't have to actively maintain it (but you will). Of course, that's not really applicable to this very PR, and I'm not sure whether it can be achieved at all. But see it as the general answer I tried to give here.

**Comment by DavideFioravanti** [comment]
Thank you for the complete explanation and, again, for your time. Surely your answer will be helpful, at least for me, in the future. :)
Now I understand better the whole process. Probably, I underestimated the "maintaining time" of this feature. 

Just to clarify, I didn't create this PR neither I was involved in the development. I am just an user/developer who tested this feature. :)

**Comment by CodeFetch** [comment]
@adschm Should be part of a wiki entry...

**Comment by adschm** [comment]
> @adschm Should be part of a wiki entry...

?

**Comment by dengqf6** [comment]
@Vladdrako R3G's 2.4 GHz WiFi chip is MT7603, which does not support VHT.

**Comment by CodeFetch** [comment]
@adschm Your explanation is very insightful and should be part of a wiki entry instead of getting lost here.

**Comment by howl** [comment]
@adschm I have read this but let me put an example, arv7519rw22, has been at least three pull request from different people to bring uboot. I like your statement that if in x time there has been no activity then there is going no activity in no time plus no time, well, that could be in both sides. I know, this is done by most people in spare time but, some pull requests hasn't take interest by the only people that could merge it till, then you act saying that there is no response from the author. That only kills the interest some people could have, and in fact, search for arv7519rw22, myself argued that code worked and the only missing thing was a brnboot uboot version.

Actually I have redone that pull request opened at least two times, with uboot asc, brn and nor versions, also a pr renabling arv7519rw22, but it doesn't get interest at all.

Now take the maintain issue, in this case is just a "receipe" to make uboot work in that device where the characteristics are defined, if there is a need of change is going to be a general change for all the boards that are in uboot-lantiq.

Let me say that I understand all you say, but don't hide that there is also some percentage of randomness where contributions are taking into consideration.

**Comment by ayyyuki1** [comment]
This works well with a lot of snapdragon phones, 1x1 20 from 72 becomes 96, 1x1 40 from 150 becomes 200.

This also works with ipq4019 and iwlwifi (ax200 AP mode). it also enables mu-mimo on the 2.4GHz radio of ipq4019 which is a bonus.

@dengqf6 is there a way to enable this but on supplicant? that way i can try to make two ipq4019 devices talk to each other over 2.4GHz (AP+4addr & STA+4addr) and have them communicate at 2x2 40 (400Mbps)?

**Comment by Ansuel** [comment]
The main issue of this is the compatibility regression of some device that
may chase to work with vht on 2.4.

Il Sab 17 Set 2022, 03:48 Richard Huynh ***@***.***> ha
scritto:

**Comment by dhewg** [comment]
I rebased the patches, see https://github.com/dhewg/openwrt/commits/vht

**Pull request closed without merging**